dpsim/src/MNASolverDirect.cpp

```diff
@@ -0,0 +1,393 @@
+/* Copyright 2017-2021 Institute for Automation of Complex Power Systems,
+ *                     EONERC, RWTH Aachen University
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ *********************************************************************************/
+
+#include <dpsim/MNASolverDirect.h>
+#include <dpsim/SequentialScheduler.h>
+
+using namespace DPsim;
+using namespace CPS;
+
+namespace DPsim {
+
+
+template <typename VarType>
+MnaSolverDirect<VarType>::MnaSolverDirect(String name, CPS::Domain domain, CPS::Logger::Level logLevel) :	MnaSolver<VarType>(name, domain, logLevel) {
+	implementationInUse = DirectLinearSolverImpl::SparseLU;
+}
+
+
+template <typename VarType>
+void MnaSolverDirect<VarType>::switchedMatrixEmpty(std::size_t index) {
+	mSwitchedMatrices[std::bitset<SWITCH_NUM>(index)][0].setZero();
+}
+
+template <typename VarType>
+void MnaSolverDirect<VarType>::switchedMatrixEmpty(std::size_t swIdx, Int freqIdx) {
+	mSwitchedMatrices[std::bitset<SWITCH_NUM>(swIdx)][freqIdx].setZero();
+}
+
+template <typename VarType>
+void MnaSolverDirect<VarType>::switchedMatrixStamp(std::size_t index, std::vector<std::shared_ptr<CPS::MNAInterface>>& comp)
+{
+	auto bit = std::bitset<SWITCH_NUM>(index);
+	auto& sys = mSwitchedMatrices[bit][0];
+	for (auto component : comp) {
+		component->mnaApplySparseSystemMatrixStamp(sys);
+	}
+	for (UInt i = 0; i < mSwitches.size(); ++i)
+		mSwitches[i]->mnaApplySwitchSparseSystemMatrixStamp(bit[i], sys, 0);
+
+	// Compute LU-factorization for system matrix
+	mDirectLinearSolvers[bit][0]->preprocessing(sys, mListVariableSystemMatrixEntries);
+
+	auto start = std::chrono::steady_clock::now();
+	mDirectLinearSolvers[bit][0]->factorize(sys);
+	auto end = std::chrono::steady_clock::now();
+	std::chrono::duration<double> diff = end-start;
+	mFactorizeTimes.push_back(diff.count());
```

Comment:
I think for performance and storage saving, the determination of `mFactorizeTimes` (and also `mSolveTimes` and `mRecomputationTimes`) should be configurable and switched off per default. This should be done via the simulation class, similarly to the method `logStepTimes` in the simulation class. You can take care of it in the branch `profiling-based-optimisation`.

Reply:
removed from this branch entirely for now